### ISSUE TITLE
Type literal subtype info discarded when type variable is nested inside enclosing container

### ISSUE THREAD

**Issue by colehaus** [microsoft/pyright#9173]
Under pyright 1.1.383, the following code:
```python
from __future__ import annotations

from typing import Generic, Literal, TypeVar

A = TypeVar("A")


class Wrapper(Generic[A]): ...


def subplots(nrows: A) -> Wrapper[A]: ...
def subplots_(nrows: A) -> tuple[int, Wrapper[A]]: ...


def foo():
    x: Wrapper[Literal[2]]
    x = subplots(2)
    z: Wrapper[Literal[2]]
    y, z = subplots_(2) # Error here
    # Just to silence the unused variable warning
    print(x, y, z)
```
produces this error:
```
pyright.py:19:12 - error: Type "Wrapper[int]" is not assignable to declared type "Wrapper[Literal[2]]"
    "Wrapper[int]" is not assignable to "Wrapper[Literal[2]]"
      Type parameter "A@Wrapper" is invariant, but "int" is not the same as "Literal[2]" (reportAssignmentType)
```
I would expect the second call to type check just like the first one. Indeed this code does type check in mypy 1.11.2. But it seems that pyright is losing the literal info in the second case when the `Wrapper` is enclosed in a tuple.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

When you assign to a variable like `x` that has a declared type, pyright uses bidirectional type inference to infer the type. When you assign a value to a tuple or list expression (like `y, z` or `[y, z]`), that expression has no declared type, so bidirectional type inference doesn't apply.

If you want pyright to use bidirectional type inference in that case, you'd need to declare an intermediate variable whose type is a tuple.

```python
    temp: tuple[int, Wrapper[Literal[2]]]
    temp = subplots_(2)
    y, z = temp
```
